### DISCUSSION TITLE
Chrome notification popups

### DISCUSSION THREAD

**Discussion by tv42** [niri-wm/niri#425] [Q&A]
Niri 0.1.6 behavior of Chrome browser notifications from e.g. Slack notifications is very disruptive and hard to work with. Right now, every notification creates a new column to the right, with a new window showing the message there. That window steals the focus (interrupting my typing), and needs to be closed with the tiny [x] in the corner (no ESC etc).

You can test this at https://www.bennish.net/web-notifications.html

Anyone have a good idea of how to improve on this?

It seems Chrome developers have completely abandoned using native desktop notifications. (https://issues.chromium.org/issues/41496153 and https://www.reddit.com/r/chrome/comments/t551ib/comment/hz84t81/?utm_source=reddit&utm_medium=web2x&context=3 and https://www.reddit.com/r/chrome/comments/t5o39l/how_to_turn_chrome_native_notifications_back_on/ and https://issues.chromium.org/issues/41292312)

Window rules, maybe? `niri msg -j focused-window` tells me the popup doesn't really identify itself well: `{"title":"","app_id":""}`, so this seems impossible.

Firefox uses desktop notifications for popups, and is much more pleasant with niri.

See also:

- Floating windows: https://github.com/YaLTeR/niri/issues/122

**Comment by YaLTeR** [comment]
Huh. How does it work on other WMs? I'm guessing they put it into the floating layer based on non-resizability, but you still have the manually close issue?

**Reply by tv42** [reply]
They float in the right lower corner of the display, and it's up to Chrome to dismiss them at some point / user clicking X. They do not steal focus or cause layout shifts. I don't know what property in particular makes them float & position.

**Reply by YaLTeR** [reply]
This is in Wayland Chrome, or in X11 Chrome?

**Reply by tv42** [reply]
I believe I've seen that behavior in Wayland Chrome on Sway, previously -- but if it somehow failed to launch in Wayland mode and went back to XWayland, I wouldn't have noticed.

**Comment by Airradda** [comment]
Was a solution to this ever found or at least a stopgap? I don't know if it'd be possible to just watch for a blank title, blank app ID window being spawned by the same process ID as the browser. For me at least, the minimum viable solution would just be able to spawn it floating so that it doesn't have such a jarring effect on the workspace.

```
  Title: ""
  App ID: ""
  Is floating: no
  PID: 3445380
```
```
Window ID 383:
  Title: "HTML5 Web Notifications Test - Bennish.net - Helium"
  App ID: "helium"
  Is floating: no
  PID: 3445380
```

**Comment by derat** [comment]
https://issues.chromium.org/issues/517327175 tracks the missing titles and app IDs (no action there from the Chrome side yet).

**Comment by derat** [comment]
No guarantees that it'll work for everyone, but having this in my config is working so far:

```kdl
window-rule {
    match app-id="^$" title="^$"
    open-floating true
    open-focused false
    default-floating-position x=4 y=4 relative-to="bottom-right"
}
```

Luckily Chrome seems to be the only app I use that opens windows with missing app IDs and titles.

**Reply by kode54** [reply]
Thanks, I was missing the `open-focused` setting, so they were occasionally stealing input focus.

As per recent testing as well, disabling the built-in notifications and forcing system notification use appears to no longer be an option.

I am also adding a commit to work around a possible issue and support setting the `suspended` state on xdg_toplevel, in case it will fix an issue where notifications opened during a display lock and power down become unresponsive until another notification comes in while the screen is active again. At one point I found it got so bad that the windows were unresponsive even after Chrome was terminated, and I had to restart the session.

**Reply by derat** [reply]
After I installed and ran [mako](https://github.com/emersion/mako/), Chrome started using it to display notifications rather than opening them in hard-to-identify `app-id`-less windows. This is still working well for me with `google-chrome-stable` 150.0.7871.114-1 on Debian, and I no longer need to use the window rule described above (although I still need similar rules for things like YouTube and Google Meet picture-in-picture windows opened by Chrome).

**Reply by kode54** [reply]
Actually, I just noticed, it is indeed using my Noctalia notifier windows. However, Noctalia needed the added benefit I just committed a day ago: When the desktop is locked, or the monitors are off, or both, it will now set the Suspended state on xdg toplevel windows that subscribe to v6 of the protocol, so notifications should be withheld or suspended while the display is locked, instead of getting stuck onscreen until the next notification comes in.
